### DISCUSSION TITLE
Quic Handshake Breakdown

### DISCUSSION THREAD

**Discussion by btoms20** [swift-quic/swift-quic#3] [General]
Hey, I came across this repo a few weeks ago and decided to try and jump in and help get Quic running in swift-nio.

### Prerequisites:
- Client Configures TLS & binds to UDP port (assume TLS 1.3)
- Server Configures TLS & binds to UDP port (assume TLS 1.3)

### Client Generates an `InitialPacket`:
- Ask TLS for Client Hello.
- Generate a Destination ConnectionID (`DCID`) (random bytes between 4 and 20 bytes long)
- Inject Quic Extension Params into TLS Client Hello.
- Wrap TLS Client Hello in Quick Crypto Frame.
- Create a Quic `Initial Packet` containing...
    - The header byte `0b11000001`
    - ```
      +------------+--------------+------------+------------+----------------------+
      | Form       | Fixed Bit    | Type       | Reserved   | Packet Number Length |
      +------------+--------------+------------+------------+----------------------+
      | [1]1000001 | 1[1]000001   | 11[00]0001 | 1100[00]01 | 110000[01]           |
      +------------+--------------+------------+------------+----------------------+
      | long       | Optional(64) | initial    | Valid      | 2                    |
      +------------+--------------+------------+------------+----------------------+
      ```
    - The 4 byte Quick Version code
    - The `DCID` generated above
    - A randomly selected Source ConnectionID (`SCID`) (between 4 and 20 bytes)
    - A zero length Token byte (0x00)
    - The payload length (in 2 byte VarInt format) 
    - The payload itself consists of the above Crypto Frame and sufficient zero byte padding.
- Generate Client Initial Key Set (`traffic`, `iv`, and `hp` keys) using the Quic initial secrets and the `DCID` generated by the client.
- Encrypt Initial Packet payload...
   - Initial Packets use the `TLS_AES_128_GCM_SHA256` cipher suite.
   - The `AES-GCM` params are as follows...
      -  The symmetric key is the client `traffic` key.
      -  The `nonce` is generated from OR(^)ing the client iv and the packet number [0x0000 in this case]).
      -  The unprotected `header` as the Authenticated Data.
   - The payload is the Ciphertext and the 16 byte Tag.
- Apply Header Protection to the Initial Packets Header...
   - `sample` 16 bytes of the encrypted payload starting at an offset of 4 - packetNumberLength  ( 2 in this case ).
      - 4 bytes is the maximum packet number length in bytes that quic allows.
      - the initial packet uses a 2 byte length packet number of 0.
      - so the `sample` offset is 4 - 2 = 2.
   - Generate a header protection mask
      - Using `AES-ECB` with the following params...
         - The symmetric key is the clients `hp` key .
      - The `mask` is the result of encrypting the `sample`.
   - Apply the mask to the last 4 bits of the first byte of the header with the first byte of the mask.
      - `header[0] ^= mask[0] & 0x0f // long headers only`
   -  Apply the mask to the packet number bytes.
      - `header[pnOffset...] ^= mask[1...]`
      - In this instance our packet number is 2 bytes so we only use the first 3 bytes of the mask.
- Initial Packet now consists of a Protected Header and an Encrypted Payload.
- Wrap this packet in an UDP AddressedEnvelope and send it off to the Server!


### Server decrypts the clients `InitialPacket`:
 - Generate Client Initial Key Set (`traffic`, `iv`, and `hp` keys) using the Quic initial secrets and the `DCID` generated by the client.
 - Using these keys the server...
    - Removes the Header Protection.
    - Decrypts the `InitialPacket`'s payload (using the unprotected header as the Authenticated Data).
 - Removes the Padding from the `InitialPacket`'s payload.
 - Consumes the Crypto Frame...
    - Removes the Crypto Frame header.
    - Reads the Quic Extension Parameters.
    - Applies the TLS Client Hello Header bytes.
    - Feed the Client Hello into the installed TLS instance.
 - TLS Responds with...
    - A Server Hello message.
    - Derived Handshake Secrets. 
    - Derived Traffic Secrets.
    - A Server Handshake Message.
 - Repeat the steps above to prepare the Server's `InitialPacket` (using the server's keys).
    - ACK Frame indicating we received the client's `InitialPacket`.
    - Crypto Frame containing the TLS Server Hello.
- Generate the Handshake keys using the Handshake Secret TLS derived. 
- Then construct the server's `HandshakePacket` using the Handshake Keys and the agreed upon Cipher Suite.
...


### Where I'm at...
So far I'm able to generate a client InitialPacket that elicits a response from a server that contains an `InitialPacket` and a `HandshakePacket` in the response. I'm able to remove header protection and decrypt the server's `InitialPacket`. I feed the TLS Server Hello into the installed TLS instance and have TLS share the derived handshake secrets via the TLS Key Logger callback. The negotiated Cipher Suite is `TLS_AES_128_GCM_SHA256`. I generate both the client and server handshake key rings using the `CLIENT_HANDSHAKE_TRAFFIC_SECRET` and `SERVER_HANDSHAKE_TRAFFIC_SECRET` respectively, just as I do with the `Initial Client Secret` and `Initial Server Secret` using the same HKDF_Expand_Label function that generates a valid Initial key ring. But I fail at removing header protection and therefor decrypting the Server's handshake packet. 

Any idea where I might be going wrong?

### Update 1
Tampering with TLS messages in flight (obviously not okay)

### Update 2
The invalid signature was the result of incorrectly parsing a VarInt

### Update 3
Handshake accomplished 🤝

**Comment by btoms20** [comment]
Kinda silly mistake here... I was tampering with the SSL handshake messages in flight (adding the Quic transport params) which caused the TLS message hashing (to prevent message tampering) to work as designed and create invalid handshake and application secrets resulting in a failed handshake. 

Lessons learned: Don't tamper with in flight TLS handshake messages haha 🙈

**Comment by kennethlaskoski** [comment]
Hi Brandon,
thank you very much for your interest and please forgive me for this late and shallow response. Truth is that I'm writing nothing but an ACK, as a token of respect.

**Reply by btoms20** [reply]
Not sure where you are regarding the quic implementation, but I'm able to decrypt and process handshake messages now and I'm in the process of trying to get past an Invalid Cert / Signature error. I was going to wait until I had the handshake completing successfully against a vanilla [quic-go](https://github.com/quic-go/quic-go) server before posting any code to this repo but if you think some WIP code would be useful maybe we can open a dev branch and iterate on some code together. If not, no worries.

**Reply by kennethlaskoski** [reply]
Again, I must apologize for the late response. I created a develop branch and invited you to collaborate. For sure, I welcome any help and would love to work with you. I've managed to allocate some time and hope this project gains momentum.

**Reply by btoms20** [reply]
Sounds good. I'll spend today/tomorrow cleaning up what I have written so far and start pushing code to the dev branch soon. Thanks for the invite.

**Reply by kennethlaskoski** [reply]
Awesome! Welcome aboard!

**Reply by btoms20** [reply]
I know it's a been a little longer than originally anticipated, but I just pushed a bunch of WIP code to the dev branch. 

Thought I'd update this thread in case anyone comes across it.

**Reply by kennethlaskoski** [reply]
Awesome! Flipping awesome!
I transferred the repo to an organization and added you as owner. I'm working on a minimal CI/CD workflow and the general scaffolding of the repository. After I manage to setup basic automation to things like PR's and reviews, I will look on your code so we can discuss scope and structure.
THANKS for your help!

**Comment by kennethlaskoski** [comment]
I've started working on my own handshake implementation. I know this is redundant and apparently a waste of time given that you already have it working, but the idea is to give us a not so abstract basis to discuss things like design and overall approach to the problem. Besides, I have to get (re) acquainted with my own code.

**Reply by btoms20** [reply]
Sounds great! Having multiple implementations should help us land on a better final design!
